scanner/src/main/kotlin/ScannerCriteria.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2020 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner
+
+import com.vdurmont.semver4j.Semver
+
+import java.util.regex.Pattern
+
+import org.ossreviewtoolkit.model.ScannerDetails
+
+/**
+ * Definition of a predicate to check whether the configuration of a scanner is compatible with the requirements
+ * specified by a [ScannerCriteria] object.
+ *
+ * When testing whether a scan result is compatible with specific criteria this function is invoked on the
+ * scanner configuration data stored in the result. By having different, scanner-specific mather functions, this
+ * compatibility check can be made very flexible.
+ *
+ * TODO Switch to a more advanced type than String to represent the scanner configuration.
+ */
+typealias ScannerConfigMatcher = (String) -> Boolean
+
+/**
+ * A data class defining selection criteria for scanners.
+ *
+ * An instance of this class is passed to a [ScanResultsStorage] to define the criteria a scan result must match,
+ * so that it can be used as a replacement for a result produced by an actual scanner. A scanner implementation
+ * creates a criteria object with its exact properties. Users can override some or all of these properties to
+ * state the criteria under which results from a storage are acceptable even if they deviate from the exact
+ * properties of the scanner. That way it can be configured for instance, that results produced by an older
+ * version of the scanner can be used.
+ */
+data class ScannerCriteria(
+    /**
+     * Criterion to match the scanner name. This can be an exact scanner name, but this property is actually a
+     * regular expression. So it is possible select multiple scanners using an alternative ('|') expression or
+     * an arbitrary one using a wildcard ('.*').
+     */
+    val regScannerName: String,
+
+    /**
+     * Criterion to match for the minimum scanner version. Results are accepted if they are produced from scanners
+     * with at least this version.
+     */
+    val minScannerVersion: Semver,
+
+    /**
+     * Criterion to match for the maximum scanner version. Results are accepted if they are produced from scanners
+     * with a version lower than this one. (This bound of the version range is excluding.)
+     */
+    val maxScannerVersion: Semver,
```

Comment:
There is probably no solution that supports all use cases equally well. I took this approach with the excluding upper bound because this is the way OSGi handles version ranges. It has the advantage that it works well with semantic versioning. You can conveniently specify that you want a semantically compatible version greater or equal than a specific version. For instance, set the minimum version to 1.5.3 and the maximum version to 2.0.0. Then all 1.x versions >= 1.5.3 are accepted.

Reply:
Agreed. But how about renaming those properties to just `minVersion` and `maxVersion` to make it a bit shorter? I think it's obvious that the scanner version is meant since they are properties of `ScannerCriteria`.